### ISSUE TITLE
Zero dice score for missing label in ground truth.

### ISSUE THREAD

**Issue by GYDDHPY** [MIC-DKFZ/nnUNet#1159]
I have a problem relevant to loss compution.

nnUNet uses $SoftDiceLoss$ as part of loss function, which computes DICE score by using $get\\_tp\\_fp\\_fn\\_tn$ function.  

The fuction first encode the groud truth into one-hot label. However, if some class is not shown in the image, which lead to label missing, the one-hot encoding may produce zeros for the channel corresponding to that class. This may cause zero tp in  $get\\_tp\\_fp\\_fn\\_tn$ function and finally a zero dice score for that class.  

The $SoftDiceLoss$ function calculates an average dice score for all classes(maybe exclude background according to settings). It seems that a zero value used in mean calculation may lead to an lower value.  

Will the missing label and zero dice score affect the final loss calculation?    

Is it necessary for us to deel with the effect of missing label in some images caused by uncontrollable resons?

https://github.com/MIC-DKFZ/nnUNet/blob/fd58e25a304e2ab0cd4c16d2c79505c2882e3593/nnunet/training/loss_functions/dice_loss.py#L100-L153  

https://github.com/MIC-DKFZ/nnUNet/blob/fd58e25a304e2ab0cd4c16d2c79505c2882e3593/nnunet/training/loss_functions/dice_loss.py#L178-L190

**Comment by FabianIsensee** [comment]
Hey, this is not a problem at all: For each sample, all classes contribute equally to the loss. If a class is absent in the gt, its contribution is 0 (for the Dice, the CE still penalizes false positives as it should). We should not reweigh the impact each class has depending on whether other classes are there or not

**Comment by GYDDHPY** [comment]
> Hey, this is not a problem at all: For each sample, all classes contribute equally to the loss. If a class is absent in the gt, its contribution is 0 (for the Dice, the CE still penalizes false positives as it should). We should not reweigh the impact each class has depending on whether other classes are there or not

Thank you for your quick reply.  I seem to understand your explanation. The loss function of nnUNet can deel with the simultaneous missing of target and corresponding label as the CE loss exists. 

What if we just don't label the shown target/foreground in the image?

I'm working on a dataset which contains images from different sources. Tha annotation standard of images are different.  

For example, there are two images, A and B, which both contain kidney. However,  due to the difference of annotation requirements, the kidney in image A is annotated but is not labeled in image B.  

A trained model is probably be able to predict kidney part correctly in image B. But due to the missing of the kidney label, the predition may be regard as false positive area.  

What should we do under such situation?    

Looking forward to your reply.

**Comment by FabianIsensee** [comment]
> What if we just don't label the shown target/foreground in the image?

Then nnU-Net will probably predict it and will get a bunch of FP in the Dice loss thus suppressing the prediction. NEVER not label something (because not labeling something equals to labeling it as background) because this will destroy your training by sending conflicting information to nnU-net

This situation can not really be resolved in a straightforward manner. If you know what organs are missing in what images you can integrate that into the training and make sure no losses are propagated from these images and organs, effectively ignoring them. But this is quite some work in the code to get it right - we did this internally (but cannot share it at this point)

**Comment by GYDDHPY** [comment]
Thank you for your reply. Your answer really help me a lot.
